### ISSUE TITLE
Inconsistencies with respect to the Polygon `within` and `intersection` methods.  

### ISSUE THREAD

**Issue by RavicharanN** [Toblerity/Shapely#966]
## Expected behavior and actual behavior.

I've been trying to use the `intersection` method to compute the intersection region of the 2 given `shapely.geometry.Polygon` instances. I was initially trying to verify the intersection results (let's call it `res`) by checking if a point I know that lies in the result using the `contains` method. 

I was then trying to dig into the source of this issue and I found that for a few cases the intersection result `res` when passed through `poly.contains(res)`, the method returns `False` where ideally it is supposed to return a `True` value as the intersection of the polygon is supposed to be a part both the Polygons. 

## Steps to reproduce the problem.
* Adding it in the comments below as I was mainly dealing with huge GeoJSON data.

## Operating system

Mac OS X Mojave, Ubuntu 18.04

## Shapely version and provenance

Shapely - 1.7.0

**Comment by RavicharanN** [comment]
### One of the example where the `contains` method is showing an anomaly so that the issue could be reproduced. 

```
poly1 = [[ 30.42974008  -0.24756188]
 [ 29.91109462   5.43627133]
 [ 28.37563434  10.93774757]
 [ 25.80548834  16.06938069]
 [ 22.19049567  20.63117008]
 [ 17.56078374  24.41219929]
 [ 12.02951807  27.20325383]
 [  5.82949138  28.8230354 ]
 [ -0.68522393  29.15337142]
 [ -7.09298153  28.16884024]
 [-12.99588514  25.94385329]
 [-18.0973928   22.63263673]
 [-22.22765217  18.4345156 ]
 [-25.32054512  13.56197731]
 [-27.37099805   8.22126498]
 [-28.39622029   2.60581202]
 [-28.41011799  -3.10155054]
 [-27.41121362  -8.72185061]
 [-25.38252573 -14.07185509]
 [-22.3035231  -18.95722093]
 [-18.17596204 -23.17012797]
 [-13.063555   -26.49576614]
 [ -7.13621672 -28.73197428]
 [ -0.69441037 -29.7218974 ]
 [  5.85649899 -29.38972191]
 [ 12.08651633 -27.76135821]
 [ 17.63567751 -24.95711804]
 [ 22.26930242 -21.16115183]
 [ 25.87561275 -16.58531145]
 [ 28.42750374 -11.44246678]
 [ 29.93852273  -5.93384244]
 [ 30.42974008  -0.24756188]]

poly2 = [[ 30.58928692   0.11904601]
 [ 30.08501649   5.90784013]
 [ 28.54806008  11.51653383]
 [ 25.95465129  16.75525535]
 [ 22.28686886  21.41990216]
 [ 17.5672003   25.29316257]
 [ 11.90527713  28.15725657]
 [  5.5394935   29.82166622]
 [ -1.15897954  30.16132395]
 [ -7.74426905  29.14923467]
 [-13.79680948  26.86432561]
 [-19.00864224  23.46930668]
 [-23.20945972  19.17269005]
 [-26.3393476   14.19453155]
 [-28.40130061   8.74627802]
 [-29.41949657   3.02451578]
 [-29.41252845  -2.78611733]
 [-28.38112259  -8.50546207]
 [-26.30820553 -13.94907376]
 [-23.17126308 -18.92080813]
 [-18.96899437 -23.20999584]
 [-13.76258122 -26.59774332]
 [ -7.72235903 -28.87697659]
 [ -1.15431991 -29.88633284]
 [  5.52579877 -29.54760791]
 [ 11.87641695 -27.88754453]
 [ 17.52936163 -25.03010996]
 [ 22.24714879 -21.16436759]
 [ 25.91938475 -16.50676856]
 [ 28.52201869 -11.27365418]
 [ 30.07126069  -5.66852797]
 [ 30.58928692   0.11904601]]

shapely_poly1 = shapely.geometry.Polygon(list(poly1))
shapely_poly2 = shapely.geometry.Polygon(list(poly2))
res = shapely_poly1.intersection(shapely_poly2)
```

In the above example `shapely_poly1.contains(res)` returns `False` but the `res` is however an instance of the `Polygon` which contains the intersection data and is not empty.

**Comment by mwtoews** [comment]
I cant reproduce the issue:
```python
from shapely import wkb
p1 = wkb.loads('0103000000010000002000000063612572036E3E401C0B2B911BB0CFBF28763C7F3DE93D4034215AE9BDBE1540F78D759229603C40611B127320E02540FD9BDD7B34CE39401F86D2EEC2113040FBAD0053C43036404FD0C35C94A13440B76DEF858F8F314003FC85E485693840039512FE1C0F284002A3687108343B40D8A73530665117404573AE72B2D23C40C62B39BC5AEDE5BF840A715943273D409DEDD98C365F1CC07E0C2D1D392B3C40A9BE35A8E4FD29C0A2C5845EA0F13940AADD0ABCEE1832C03099117BF4A136407003A169473A36C0119A136A3C6F3240DA4AB73E0F5239C08E166F7DBB1F2B4039A16BBAF95E3BC040B3B9A449712040C78F63B16E653CC0BC63EBF8B3D804401F8D1A7EFD683CC07487C1BAF9CF08C0DD91B94B45693BC066183567967121C0813ACC34ED6139C0C833BB30CA242CC09E149CB0B34D36C08E654D6E0CF532C0302327D90B2D32C0ED48B3818D2B37C09BFEEC478A202AC0A0C39D87EA7E3AC0084656657C8B1CC0F71C9AAA62BB3CC081E7A4189C38E6BF43119C44CEB83DC09A6D3C120E6D174012FCA9D0C4633DC003D74FDE4B2C2840B97D245FE8C23BC0B240E4C2BBA231402D3618B005F538C037A2DE00F1443640628A0F3F412935C0EB353D2828E039406EB99FF8D69530C0BA3796E2706D3C4036557B018BE226C013E78F0643F03D40D9AA4D3141BC17C063612572036E3E401C0B2B911BB0CFBF', hex=True)
p2 = wkb.loads('010300000001000000200000004E5CF181DB963E408556AB9FCC79BE3FB42B04A4C3153E40B462D1D7A0A11740D8D757AA4D8C3C405841461F770827405CA2E50664F439400861246A58C13040D8F0D33C7049364042B83CB57E6B354009C8F20934913140B28FC2B30C4B39405F55E67B80CF27402C3B71F741283C40499EEBFB70281640C120A7B758D23D40AD8750212E8BF2BFC14FC1864C293E4030B374AA21FA1EC050144B3E34263D404C7C8369F7972BC0960B747144DD3A40082ABA60360233C0F3648A7B24783740B53AF7269F3537C03818456A352C3340E1F9FB7BDF563AC09494AAA399632C409AD003A3BB663CC04D3F1327187E214026EF9020646B3DC0FDAF4A5435320840696BE9769B693DC0C959F9E1F74906C01BD1034091613CC07409A8ECCB0221C0DDCBBF8EE64E3AC02D62F1FEECE52BC0BB9CAFE5D72B37C0AD3DE414BAEB32C07B28D90310F832C097189149C23537C037E3B00B71862BC06ECDCAB405993AC0F847E715B2E31EC05968AD8981E03CC0882569271878F2BFBA2B81B5E6E23DC046B525FE6A1A16401D7B3008308C3DC0F4D1F3B8B9C02740FC174A1E36E33BC0759B683E84873140DF7B4D49B50739C0C54BA224453F364085928FFE132A35C0EEB089CC5CEB39404DD79795BB8130C065735104A3853C40030793661C8C26C0570AFD233E123E40F1719E9892AC16C04E5CF181DB963E408556AB9FCC79BE3F', hex=True)
res = p1.intersection(p2)
assert p1.contains(res)
assert p2.contains(res)
```

**Comment by RavicharanN** [comment]
@mwtoews I was using the `shapely.geometry.Polygon` methods for the same. As I told that I was maily dealing with GeoJSON data, here is the script that would help you reproduce the issue.

```
import shapely
import geog
import numpy as np
angles = np.linspace(0, 360, 32)
p = shapely.geometry.Point((0.9502405524253845, -0.2843794822692871))
d = 3277937.1689927974
poly1 = geog.propagate(p, angles, d)
p = shapely.geometry.Point((0.5228247046470642, 0.1375546455383301))
d = 3343230.099535113
poly2 = geog.propagate(p, angles, d)
shapely_poly1 = shapely.geometry.Polygon(list(poly1))
shapely_poly2 = shapely.geometry.Polygon(list(poly2))
>>> shapely_poly1.intersection(shapely_poly2)
>>> shapely_poly1.contains(res))
```
I was using the `geog.propogate` to produce the desired Polygon and the result returned by that method is a NumPy array. So that is being converted and passed into a `shapely` methods accordingly. 

I have verified that that generated results for the intersection remain the same irrespective of the method used to compute `poly1` and `poly2`.

**Comment by mwtoews** [comment]
Fundamentally, this is a floating point precision issue, and is not entirely a shapely issue. Yes, with a perfect precision model I would expect the result of an intersection of two geometries to be contained by each geometry. However, this cannot always be safely represented with a double precision model used by Shapely. JTS fails with this specific example too, except when the precision model is changed to a fixed with scale 1000.0.

You can apply a negative buffer with a very small amount, such as a number near [machine epsilon](https://en.wikipedia.org/wiki/Machine_epsilon) of the fraction of square root of area of the intersected area, which should return the desired result:
```python
from math import sqrt
eps = 1e-14
res2 = res.buffer(-sqrt(res.area) * eps)
assert shapely_poly1.contains(res2)
assert shapely_poly2.contains(res2)
```
note that these two geometries are nearly the same:
```python
res2.area  # 2742.7512489986843
res.area  # 2742.7512489986875
assert abs(res2.area - res.area) / res.area < 1e-12
```

**Comment by RavicharanN** [comment]
Thanks for clarifying!
